compose/rest/record.go

```diff
@@ -126,51 +128,163 @@ func (ctrl *Record) Read(ctx context.Context, r *request.RecordRead) (interface{
+							Name:  br.RefField,
+							Value: strconv.FormatUint(m.ID, 10),
+						})
+						return r, nil
+					}
+				}
+
+				batch = append(batch, b)
+			}
+		}
+	}
+
+	rs, err := ctrl.record.With(ctx).Operate(batch...)
 
 	if rve, is := err.(*types.RecordValueErrorSet); is && !rve.IsValid() {
 		return ctrl.handleValidationError(rve), nil
 	}
 
-	return ctrl.makePayload(ctx, m, record, err)
+	if rc.Records == nil {
+		return ctrl.makePayload(ctx, m, rs[0], err)
+	}
+	return ctrl.makeFilterPayload(ctx, m, rs, nil, err)
 }
 
-func (ctrl *Record) Update(ctx context.Context, r *request.RecordUpdate) (interface{}, error) {
+func (ctrl *Record) Update(ctx context.Context, ru *request.RecordUpdate) (interface{}, error) {
 	var (
 		m   *types.Module
 		err error
 	)
 
-	if m, err = ctrl.module.With(ctx).FindByID(r.NamespaceID, r.ModuleID); err != nil {
+	if m, err = ctrl.module.With(ctx).FindByID(ru.NamespaceID, ru.ModuleID); err != nil {
 		return nil, err
 	}
 
-	record, err := ctrl.record.With(ctx).Update(&types.Record{
-		ID:          r.RecordID,
-		NamespaceID: r.NamespaceID,
-		ModuleID:    r.ModuleID,
-		Values:      r.Values,
-	})
+	batch := make([]*types.BulkRecordOperation, 0)
+
+	// Initialize parent record for creation, if defined
+	if ru.Values != nil {
+		rr := &types.Record{
+			ID:          ru.RecordID,
+			NamespaceID: ru.NamespaceID,
+			ModuleID:    ru.ModuleID,
+			Values:      ru.Values,
+		}
+		batch = append(batch, &types.BulkRecordOperation{
+			Record:    rr,
+			Operation: types.OperationTypeUpdate,
+		})
+	}
+
+	// Initialize sub records for creation, if defined
+	if ru.Records != nil {
+		for _, br := range ru.Records {
+			for _, rr := range br.Set {
+				// Default to current module and namespace; can be overwritten
+				if rr.NamespaceID == 0 {
+					rr.NamespaceID = ru.NamespaceID
+				}
+				if rr.ModuleID == 0 {
+					rr.ModuleID = ru.ModuleID
+				}
+				b := &types.BulkRecordOperation{
+					Record:    rr,
+					Operation: types.OperationTypeUpdate,
+				}
+
+				// If no RecordID is defined, we should create it
+				if rr.ID == 0 {
+					b.Operation = types.OperationTypeCreate
+				}
+
+				// If a ref field is defined, then sub records should get their fields updated after
+				// parent record is created.
+				if br.RefField != "" {
+					if ru.Values == nil {
+						return nil, errors.New("missing parent record definition")
+					}
+					b.PreProcess = func(m, r *types.Record) (*types.Record, error) {
+						r.Values = r.Values.Set(&types.RecordValue{
+							Name:  br.RefField,
+							Value: strconv.FormatUint(m.ID, 10),
+						})
+						return r, nil
+					}
+				}
+
+				batch = append(batch, b)
+			}
+		}
+	}
+
+	rs, err := ctrl.record.With(ctx).Operate(batch...)
 
 	if rve, is := err.(*types.RecordValueErrorSet); is && !rve.IsValid() {
 		return ctrl.handleValidationError(rve), nil
 	}
 
-	return ctrl.makePayload(ctx, m, record, err)
+	if ru.Records == nil {
```

Comment:
This will hurt us. Bad.

Is there anyway to keep consistent return value?

Reply:
If we wish to return all the changed records (which makes sense), I can't think of a way to do so without breaking BC.
Not my proudest solution, but it did seem ok for now.

Because of this, I might prefer to define a new endpoint, that we can migrate to at a later point in time.

Reply:
Hmm... The situation is as follows. As is, the record is provided directly in the `response` on the very root.

If we just add `records RecordSet`, we can get a weird structure, that I don't agree on:
```
{
  response: {
    recordID: '...',
    moduleID: '...',
    values: [...],
    ...

    records: [...],
  }
}
```

Reply:
We could, however, provide either a single record (as we do now) or a list of records.
so either
```
{
  response: {
    recordID: '...',
    moduleID: '...',
    values: [...],
    ...
  }
}
```

or

```
{
  response: [{
    recordID: '...',
    moduleID: '...',
    values: [...],
  }]
}
```

But this is fairly similar to the original solution...